File: DataInjectorIOS/Example/DataInjector/helper/MockDependency.swift
//
//  MockDependency.swift
//  DataInjector example
//
//  Example helper: define a dependency for the dependency manager which can resolve itself through the example mock JSON files
//

import Foundation
import DataInjector

class MockDependency: InjectorDependency {

    // --
    // MARK: Members
    // --
    
    private var injectors: [DataInjector] = [
        SnakeToCamelCaseInjector(),
        ReplaceNullInjector()
    ]
    private let filename: String
    private var storedJson: [Any]?
    

    // --
    // MARK: Initialization
    // --
    
    init(name: String, filename: String) {
        self.filename = filename
        if filename.hasPrefix("customer") {
            injectors.append(JoinStringInjector(item: "fullName", fromItems: [ "~firstName", "~middleName", "~lastName" ], delimiter: " ", removeOriginals: true))
        }
        super.init(name: name)
    }
    

    // ---
    // MARK: Data access
    // ---
    
    override open func obtainInjectableData() -> Any? {
        return storedJson
    }
    
    
    // --
    // MARK: Resolving
    // --
    
    private func jsonData(forFilename: String) -> Data? {
        let bundle = Bundle.main
        if let path = bundle.path(forResource: "data/" + forFilename, ofType: "json") {
            return try? NSData(contentsOfFile: path, options: .mappedIfSafe) as Data
        }
        return nil
    }
    
    private func processedJsonArray(forData: Data) -> [Any]? {
        if let json = try? JSONSerialization.jsonObject(with: forData, options: .allowFragments) {
            if let jsonArray = json as? [Any] {
                var processedJsonArray: [Any] = []
                for var jsonArrayItem in jsonArray {
                    for injector in self.injectors {
                        jsonArrayItem = injector.appliedInjection(targetData: jsonArrayItem)
                    }
                    processedJsonArray.append(jsonArrayItem)
                }
                return processedJsonArray
            }
        }
        return nil
    }
    
    override open func resolve(input: [String: String], completion: @escaping (_ success: Bool) -> Void) {
        let filename = self.filename
        DispatchQueue.global(qos: .default).async {
            // First load and process the data
            var processedJsonArray: [Any]?
            if let jsonData = self.jsonData(forFilename: filename) {
                processedJsonArray = self.processedJsonArray(forData: jsonData)
            }
            
            // Then apply the changes on the main thread, wait for half a second to simulate a delay in network traffic
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5, execute: {
                if processedJsonArray != nil {
                    if processedJsonArray != nil {
                        self.storedJson = processedJsonArray
                    }
                }
                completion(processedJsonArray != nil)
            })
        }
    }

}

File: DataInjectorAndroid/DataInjectorExample/src/main/java/com/crescentflare/datainjectorexample/helper/MockDependency.java
package com.crescentflare.datainjectorexample.helper;

import android.os.Handler;

import com.crescentflare.datainjector.dependency.InjectorDependency;
import com.crescentflare.datainjector.injector.DataInjector;
import com.crescentflare.datainjector.injector.JoinStringInjector;
import com.crescentflare.datainjector.injector.ReplaceNullInjector;
import com.crescentflare.datainjector.injector.SnakeToCamelCaseInjector;
import com.crescentflare.datainjectorexample.ExampleApplication;
import com.crescentflare.datainjectorexample.R;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Define a dependency for the dependency manager which can resolve itself through the example mock JSON files
 */
public class MockDependency extends InjectorDependency
{
    // ---
    // Members
    // ---

    private List<DataInjector> injectors = new ArrayList<>();
    private List<Object> storedJson = null;
    private int rawResourceId;


    // ---
    // Initialization
    // ---

    public MockDependency(String name, int rawResource)
    {
        // Add basic injectors
        super(name);
        injectors.addAll(Arrays.asList(
                new SnakeToCamelCaseInjector(),
                new ReplaceNullInjector()
        ));

        // Store loading resource and add extra injector for the customer list
        rawResourceId = rawResource;
        if (rawResourceId == R.raw.customer_list)
        {
            injectors.add(new JoinStringInjector("fullName", Arrays.asList("~firstName", "~middleName", "~lastName"), " ", true));
        }
    }


    // ---
    // Data access
    // ---

    @Override
    public Object obtainInjectableData()
    {
        return storedJson;
    }


    // ---
    // Resolving
    // ---

    @Override
    public void resolve(Map<String, String> input, final CompleteListener completeListener)
    {
        final Handler handler = new Handler();
        final int rawResourceId = this.rawResourceId;
        Thread loadThread = new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                // First load and process the data
                List<Object> processedJson = null;
                InputStream stream = ExampleApplication.context.getResources().openRawResource(rawResourceId);
                if (stream != null)
                {
                    String jsonString = readFromInputStream(stream);
                    if (jsonString != null)
                    {
                        Type type = new TypeToken<List<Object>>(){}.getType();
                        processedJson = new Gson().fromJson(jsonString, type);
                        if (processedJson != null)
                        {
                            for (Object item : processedJson)
                            {
                                for (DataInjector injector : injectors)
                                {
                                    injector.apply(item, null, null);
                                }
                            }
                        }
                    }
                }

                // Then apply the changes on the main thread, wait for half a second to simulate a delay in network traffic
                final List<Object> applyJson = processedJson;
                handler.postDelayed(new Runnable()
                {
                    @Override
                    public void run()
                    {
                        if (applyJson != null)
                        {
                            storedJson = applyJson;
                        }
                        completeListener.onResolveResult(applyJson != null);
                    }
                }, 500);
            }
        });
        loadThread.start();
    }


    // ---
    // Helper
    // ---

    private static String readFromInputStream(InputStream stream)
    {
        final int bufferSize = 1024;
        final char[] buffer = new char[bufferSize];
        final StringBuilder out = new StringBuilder();
        String result = null;
        try
        {
            Reader in = new InputStreamReader(stream, "UTF-8");
            for ( ; ; )
            {
                int rsz = in.read(buffer, 0, buffer.length);
                if (rsz < 0)
                {
                    break;
                }
                out.append(buffer, 0, rsz);
            }
            result = out.toString();
            stream.close();
        }
        catch (Exception ignored)
        {
        }
        return result;
    }
}
